Adds content locking to navigation block

draganescu opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

The navigation block, when locked, cannot restrict editing of contents (links and 
submenus). This PR fixes this.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Because it should work at least on the same feature level as a reusable block, which 
has this ability to restrict editing of contents.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Copies the same way in which the reusable blocks have the feature, thus also adding 
a new block api method `isNavigationBlock`.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->


1. Add a post
2. Insert a navigation block
3. Insert some links
4. From the more menu (three dots) choose lock
- Does "Restrict editing" appear?
6. Check Restrict editing
7. Click Apply

- Does the setting apply? You should not be able to modify the navigation block's 
contents.

## Screenshots or screencast <!-- if applicable -->



https://user-images.githubusercontent.com/107534/194833645-ab27fee8-b1d4-41d3-a993-74b9ec3f6330.mp4




## To do

- [x] Make block locking disable block reordering via list view as well
- [ ] Refactor the way we figure out entity based blocks available for locking

draganescu committed:
Adds content locking to navigation block.
Adds a new block api method `isNavigationBlock` to complement `isReusableBlock`.

draganescu committed:
Fixes locking UI on the navigation block.

talldan commented on packages/blocks/src/api/registration.js:
```diff
@@ -608,6 +608,18 @@ export function isReusableBlock( blockOrType ) {
+ *
+ * @return {boolean} Whether the given block is a navigation block.
+ */
+export function isNavigationBlock( blockOrType ) {
+	return blockOrType?.name === 'core/navigation';
+}
```
I know there's already `isTemplatePart` and `isReusableBlock`, but IMO those are a mistake. The blocks package should be agnostic to the names in the block library. I don't think more of these functions should be added.

Mamaduka replied:
I agree. Probably not worth introducing a new API method just for this check.

@draganescu, what do you think if we update edit locking checks to something like this?

```js
// A module-level const
const ALLOWS_EDIT_LOCKING = [ 'core/block', 'core/navigation' ];


// Then in `mapSelect` function
return {
  ...
  allowsEditLocking: ALLOWS_EDIT_LOCKING.includes( blockName ),
};
```

draganescu replied:
Sure thing!

draganescu replied:
Solved in [30d5d7e](https://github.com/WordPress/gutenberg/pull/44739/commits/30d5d7ed15a9a597dd2e6041cf89cf71c6cb9971)

draganescu commented:
This PR now adds locking to the navigation block that I think works like all the other blocks.

draganescu committed:
Refactors block type selection for content edit lock

- removes the newly added isNavigationBlock
- removes the dependency on isReusable from block locking UI

Co-authored-by: George Mamadashvili <240569+Mamaduka@users.noreply.github.com>

Mamaduka reviewed (approved):
Thanks for working on this, @draganescu!

I tested edit locking across browsers, and it works as expected.

Ideas for follow-up PRs. These aren't related to changes introduced here.

* Disable "Apply to all blocks inside" when only edit locking is selected.
* Disabled ability to expand blocks in the List View when the parent block has edit locking enabled.

bph commented:
Removed the label `needs dev note` re: [discussion on the Dev Note tracking issue](https://github.com/WordPress/gutenberg/issues/47771#issuecomment-1424391618)
